File: core/src/test/java/hello/core/singleton/SingletonTest.java
package hello.core.singleton;

import hello.core.AppConfig;
import hello.core.member.service.MemberService;
import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.context.annotation.AnnotationConfigApplicationContext;

public class SingletonTest {

    @Test
    @DisplayName("스프링 없는 순수한 DI 컨테이너")
    void pureContainer() {
        AppConfig appConfig = new AppConfig();
        /**
         * 1. 조회: 호출 할 때마다 생성
         */

        MemberService memberService1 = appConfig.memberService();
        MemberService memberService2 = appConfig.memberService();
        MemberService memberService3 = appConfig.memberService();

        // 참조값이 다른지 확인
        System.out.println("memeberService1 = " + memberService1);
        System.out.println("memeberService2 = " + memberService2);
        System.out.println("memeberService3 = " + memberService3);

        // memberService1 != memberService2
        Assertions.assertThat(memberService1).isNotSameAs(memberService2);
    }


    @Test
    @DisplayName("싱글톤 패턴을 적용한 객체 사용")
    void singletonServiceTest() {
        SingletonService singletonService1 = SingletonService.getInstance();
        SingletonService singletonService2 = SingletonService.getInstance();

        System.out.println("singletonService1 = " + singletonService1);
        System.out.println("singletonService2 = " + singletonService2);


        Assertions.assertThat(singletonService1).isSameAs(singletonService2);

    }

    /**
     * 스프링은 싱글톤 패턴의 단점을 모두 보완하고 싱글톤의 장점을 가져갈 수 있도록 도와준다.
     *
     */

    @Test
    @DisplayName("스프링 컨테이너와 싱글톤")
    void springContainer() {
//        AppConfig appConfig = new AppConfig();
        /**
         * 1. 조회: 호출 할 때마다 생성
         */

        AnnotationConfigApplicationContext ac = new AnnotationConfigApplicationContext(AppConfig.class);

        MemberService memberService1 = ac.getBean("memberService", MemberService.class);
        MemberService memberService2 = ac.getBean("memberService", MemberService.class);
        MemberService memberService3 = ac.getBean("memberService", MemberService.class);

        // 참조값이 다른지 확인
        System.out.println("memeberService1 = " + memberService1);
        System.out.println("memeberService2 = " + memberService2);
        System.out.println("memeberService3 = " + memberService3);

        // memberService1 != memberService2
        Assertions.assertThat(memberService1).isSameAs(memberService2);
    }


}

File: core-kotlin/src/test/kotlin/hello/corekotlin/singleton/SingletonTest.kt
package hello.corekotlin.singleton

import hello.corekotlin.AppConfig
import hello.corekotlin.member.service.MemberService
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.getBean
import org.springframework.context.annotation.AnnotationConfigApplicationContext

class SingletonTest {

    @Test
    fun `스프링 없는 순수한 DI 컨테이너`() {

        val appConfig = AppConfig()

        /**
         * 1. 조회: 호출할 마다 생성
         */

        val memberService1 = appConfig.memberService()
        val memberService2 = appConfig.memberService()
        val memberService3 = appConfig.memberService()

        // 참조값이 다른지 확인
        println("memberService1 = $memberService1")
        println("memberService2 = $memberService2")
        println("memberService3 = $memberService3")

        // memberService1 != memberService2
        Assertions.assertThat(memberService1).isNotSameAs(memberService2)
    }

    @Test
    fun `싱글톤 패턴을 적용한 객체 사용`() {

        // 코틀린의 경우 Singleton pattern을 접근할 떄 r

        val singletonService1 = SingletonService
        val singletonService2 = SingletonService

        println("singletonService1 = ${SingletonService.printMsg(singletonService1.toString())}")
        println("singletonService2 = ${SingletonService.printMsg(singletonService2.toString())}")

        Assertions.assertThat(singletonService1).isSameAs(singletonService2)
    }

    @Test
    fun `스프링 컨테이너와 싱글톤`() {

        val ac = AnnotationConfigApplicationContext(AppConfig::class.java)

        /**
         * 1. 조회: 호출할 마다 생성
         */

        val memberService1 = ac.getBean("memberService", MemberService::class.java)
        val memberService2 = ac.getBean("memberService", MemberService::class.java)
        val memberService3 = ac.getBean("memberService", MemberService::class.java)

        // 참조값이 다른지 확인
        println("memberService1 = $memberService1")
        println("memberService2 = $memberService2")
        println("memberService3 = $memberService3")

        // memberService1 != memberService2
        Assertions.assertThat(memberService1).isSameAs(memberService2)
    }

}